src/main/java/com/ably/kafka/connect/offset/OffsetRegistryService.java

```diff
@@ -0,0 +1,75 @@
+package com.ably.kafka.connect.offset;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is responsible for keeping track of the latest offset for each topic partition.
+ */
+public class OffsetRegistryService implements OffsetRegistry {
+
+    // Hashmap of TopicPartition to Offsets.
+    private final Map<TopicPartition, Long> topicPartitionToOffsetMap = new ConcurrentHashMap<>();
+
+    public OffsetRegistryService() {
+
+    }
+
+    /**
+     * Updates the topicPartitionToOffsetMap with the latest offset for each topic partition.
+     * @param records
+     */
+    @Override
+    public void updateTopicPartitionToOffsetMap(List<SinkRecord> records) {
+        for (SinkRecord record : records) {
```

Comment:
With the race condition below fixed, I think this code will be correct but not necessarily optimal. This list of records will contain *all* records submitted in one chunk, which we're hoping will be a few hundred in normal operation. There are, realistically, only a few partitions being worked on by this sink task, so in this list of a few hundred sink records, you will have many offsets per topic partition and we're really only interested in the highest offset per partition.

Updating the `ConcurrentHashMap` below is a relatively expensive operation, as it requires locking and competing with any other threads that may currently hold the lock. Hitting that lock a few hundred times just to update a small number of keys seems quite inefficient.

I think we should first filter this list to just contain the highest offset per partition, then process that filtered list with the loop below to avoid unnecessary contention for the lock on this map.